src/lib/support/PersistentArray.h

```diff
@@ -0,0 +1,316 @@
+            {
+                // Existing value, index must match
+                VerifyOrReturnError(i == index, CHIP_ERROR_DUPLICATE_KEY_ID);
+                ReturnErrorOnFailure(this->Copy(e, value));
+                return this->Save();
+            }
+        }
+
+        // New value
+        if (index < this->mCount)
+        {
+            // Override (implicit remove)
+            EntryType old_value = mEntries[index];
+            ReturnErrorOnFailure(this->Copy(mEntries[index], value));
+            ReturnErrorOnFailure(this->Save());
+            OnEntryRemoved(old_value);
+        }
+        else
+        {
+            // Insert last
+            VerifyOrReturnError(this->mCount < this->mLimit, CHIP_ERROR_INVALID_ARGUMENT);
+            ReturnErrorOnFailure(this->Copy(mEntries[this->mCount++], value));
+            ReturnErrorOnFailure(this->Save());
+        }
+        OnEntryAdded(value);
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * Stores a new entry. The Compare() method is used to ensure that the entry is unique.
+     * If the new entry matches an existing entry, do_override must be true, otherwise
+     * CHIP_ERROR_DUPLICATE_KEY_ID is returned.
+     * @param index Zero-based position within the table.
+     * @param entry Value to be store in the table.
+     * @return CHIP_NO_ERROR on success. CHIP_ERROR_DUPLICATE_KEY_ID is the entry is
+     * already stored in the table. CHIP_ERROR_INVALID_ARGUMENT if index >= limit.
+     */
+    CHIP_ERROR Set(const EntryType & value, bool do_override = true)
+    {
+        CHIP_ERROR err = this->Load();
+        VerifyOrReturnError(CHIP_ERROR_NOT_FOUND == err || CHIP_NO_ERROR == err, err);
+
+        // Check existing
+        for (size_t i = 0; i < this->mCount; ++i)
+        {
+            EntryType & e = mEntries[i];
+            if (this->Compare(value, e))
+            {
+                // Already registered
+                VerifyOrReturnError(do_override, CHIP_ERROR_DUPLICATE_KEY_ID);
+                ReturnErrorOnFailure(this->Copy(e, value));
+                return this->Save();
+            }
+        }
+        // Insert last
+        VerifyOrReturnError(this->mCount < this->mLimit, CHIP_ERROR_INVALID_LIST_LENGTH);
+        ReturnErrorOnFailure(this->Copy(mEntries[this->mCount++], value));
+        ReturnErrorOnFailure(this->Save());
+        OnEntryAdded(value);
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief Returns the stored entry at a given index.
+     * @param index Zero-based position within the table.
+     * @param entry On success, contains the value matching the given index.
+     * @return CHIP_NO_ERROR on success,
+     *         CHIP_ERROR_NOT_FOUND if index is greater than the index of the last entry on the table.
+     */
+    CHIP_ERROR Get(size_t index, EntryType & value)
+    {
+        ReturnErrorOnFailure(this->Load());
+        VerifyOrReturnError(index < this->mLimit, CHIP_ERROR_NOT_FOUND);
```

Comment:
Again, shouldn't we just ensure that mCount <= mLimit?

Reply:
We should, and we do, in the Set method. The Get method do not modify either mCount, or mLimit, so checking mCount <= mLimit here is gratuitous.